### PR TITLE
Feature/liquidate with spot sell

### PR DESCRIPTION
**Pull request by davidvuong**
- Integrates with a wsteth synthetix spot market for collateral quotes. Subsequently this means the only collateral available for margin are those issued via the spot market and susd.
- Further improvements to liquidation mechanism, introducing a rolling window for liquidation caps (https://github.com/Synthetixio/bfp-market/pull/20)
- swsteth is sold on liquidation for susd and deposited back into markets as credit for stakers/market to mint against
- More tests, more docs, more comments
- Bug fixes on reducing market size/skew liqSize during partial liquidations
- Introduced an endorsed keeper which does _not_ receive any liquidation rewards for if used to flag. It only receives a keeper fee for upkeep
- Bug fix on market-wide collateral deposit caps (https://github.com/Synthetixio/bfp-market/pull/21)
- Price divergence checks (https://github.com/Synthetixio/bfp-market/pull/22)

### CODE REVIEW

File: contracts/modules/MarginModule.sol [hunk 1]
```diff
@@ -182,33 +189,41 @@ contract MarginModule is IMarginModule {
...
+
+        Margin.CollateralType storage collateral = globalMarginConfig.supported[synthMarketId];
+        uint256 maxAllowable = collateral.maxAllowable;
 
-        Margin.CollateralType storage collateral = globalMarginConfig.supported[collateralType];
-        if (collateral.maxAllowable == 0) {
-            revert ErrorUtil.UnsupportedCollateral(collateralType);
+        // Prevent any operations if this synth isn't supported as collateral.
+        if (maxAllowable == 0) {
+            revert ErrorUtil.UnsupportedCollateral(synthMarketId);
```
Comment (davidvuong) [inline contracts/modules/MarginModule.sol:199]: Add a test in setCollateralConfiguration to see if a revert is thrown when an invalid synthMarketId is configured.

File: contracts/modules/OrderModule.sol [hunk 2]
```diff
@@ -178,21 +166,59 @@ contract OrderModule is IOrderModule {
...
+    ) external {
+        Account.loadAccountAndValidatePermission(accountId, AccountRBAC._PERPS_COMMIT_ASYNC_ORDER_PERMISSION);
+
+        PerpMarketConfiguration.GlobalData storage globalConfig = PerpMarketConfiguration.load();
+        PerpMarket.Data storage market = PerpMarket.exists(marketId);
+        validateOrderAvailability(accountId, marketId, market, globalConfig);
+        uint256 oraclePrice = market.getOraclePrice();
+
+        // TODO: Consider removing and only recomputing funding at the settlement.
+        recomputeFunding(market, oraclePrice);
```
Comment (davidvuong) [inline contracts/modules/OrderModule.sol:189]: Remove this.

File: contracts/modules/PerpAccountModule.sol [hunk 3]
```diff
@@ -24,28 +24,35 @@ contract PerpAccountModule is IPerpAccountModule {
...
         Margin.Data storage accountMargin = Margin.load(accountId, marketId);
 
-        uint256 length = globalMarginConfig.supportedAddresses.length;
-        IPerpAccountModule.DepositedCollateral[] memory collateral = new DepositedCollateral[](length);
-        address collateralType;
+        uint256 length = globalMarginConfig.supportedSynthMarketIds.length;
+        IPerpAccountModule.DepositedCollateral[] memory depositedCollaterals = new DepositedCollateral[](length);
+        uint128 synthMarketId;
+        uint256 collateralAvailable;
+        uint256 collateralPrice;
```
Comment (davidvuong) [inline contracts/modules/PerpAccountModule.sol:36]: Remove variable

File: contracts/storage/Margin.sol [hunk 4]
```diff
@@ -148,16 +181,65 @@ library Margin {
...
         for (uint256 i = 0; i < length; ) {
-            collateralType = globalMarginConfig.supportedAddresses[i];
-            available = accountMargin.collaterals[collateralType];
+            synthMarketId = globalMarginConfig.supportedSynthMarketIds[i];
+            available = accountMargin.collaterals[synthMarketId];
 
+            // Avoid changing the value to zero if already zero.
             if (available > 0) {
-                accountMargin.collaterals[collateralType] = 0;
+                accountMargin.collaterals[synthMarketId] = 0;
```
Comment (davidvuong) [inline contracts/storage/Margin.sol:194]: Remove this function and add this to the caller:


```
market.depositedCollateral[SYNTHETIX_USD_MARKET_ID] -= available;
accountMargin.collaterals[SYNTHETIX_USD_MARKET_ID] = 0;
```

and a test to verify market deposited amounts reflect a fully liquidated position. By the time we reach here, we've already flagged so there is only sUSD remaining and a full liquidation means all sUSD - fees - liqreward goes to stakers.

File: contracts/storage/Margin.sol [hunk 5]
```diff
@@ -190,51 +278,53 @@ library Margin {
...
+            synthMarketId = globalMarginConfig.supportedSynthMarketIds[i];
+            available = accountMargin.collaterals[synthMarketId];
 
-            // `INodeModule.process()` is an expensive op, skip if we can.
-            available = accountMargin.collaterals[collateralType];
+            // `getCollateralPrice()` is an expensive op, skip if we can.
             if (available > 0) {
-                price = INodeModule(globalConfig.oracleManager).process(collateral.oracleNodeId).price.toUint();
-                collateralValueUsd += available.mulDecimal(price);
+                collateralUsd += available.mulDecimal(getCollateralPrice(synthMarketId, available, globalConfig));
```
Comment (davidvuong) [inline contracts/storage/Margin.sol:293]: Don't call getCollateralPrice and use `.quoteSellExactIn(synthMarketId, available);`

File: contracts/storage/PerpMarket.sol [hunk 6]
```diff
@@ -143,6 +143,29 @@ library PerpMarket {
         self.debtCorrection = (self.debtCorrection + newCorrection - oldCorrection).to128();
     }
 
+    /**
+     * @dev Updates the `pastLiquidations` array by either appending a new timestamp or
+     */
+    function updateAccumulatedLiquidation(PerpMarket.Data storage self, uint128 liqSize) internal {
+        uint64 currentTime = block.timestamp.to64();
```
Comment (davidvuong) [inline contracts/storage/PerpMarket.sol:150]: u128 instead of 64.

File: contracts/storage/PerpMarket.sol [hunk 7]
```diff
@@ -143,6 +143,29 @@ library PerpMarket {
...
+        uint256 length = self.pastLiquidations.length;
+
+        // No prior liquidations, push new liquidation chunk.
+        if (length == 0) {
+            self.pastLiquidations.push([currentTime, uint128(liqSize)]);
```
Comment (davidvuong) [inline contracts/storage/PerpMarket.sol:155]: Remove the uint128 wrap.

File: contracts/storage/Position.sol [hunk 8]
```diff
@@ -370,12 +371,16 @@ library Position {
         uint256 marginUsd,
         uint256 price,
         PerpMarketConfiguration.Data storage marketConfig
-    ) internal view returns (uint256 healthFactor, int256 accruedFunding, int256 pnl, uint256 remainingMarginUsd) {
+    )
+        internal
+        view
+        returns (uint256 healthFactor, int256 accruedFunding, int256 unrealizedPnl, uint256 remainingMarginUsd)
```
Comment (davidvuong) [inline contracts/storage/Position.sol:377]: Rename to realizedPnl

Comment (davidvuong) [reply contracts/storage/Position.sol:377]: Renaming to pnl as there are times where the same data is used to show realized vs unrealized pnl (e.g. context of getPositionDigest will return unrealized pnl for a live position whereas on settlement, it's realized pnl).

File: contracts/interfaces/IOrderModule.sol [hunk 9]
```diff
@@ -29,14 +27,14 @@ interface IOrderModule is IBasePerpMarket {
         uint256 orderFee,
         uint256 keeperFee,
         int256 accruedFunding,
-        int256 pnl,
+        int256 unrealizedPnl,
```
[the remark this answers is not in the document]
Comment (davidvuong) [reply contracts/interfaces/IOrderModule.sol:30]: See comment below.

File: contracts/modules/MarginModule.sol (hunk 1 above)
Comment (davidvuong) [reply contracts/modules/MarginModule.sol:199]: It already reverts with `transaction reverted in contract unknown: 0x` so I've added a test to make sure this is still the case in the future.

File: contracts/storage/Margin.sol (hunk 4 above)
Comment (davidvuong) [reply contracts/storage/Margin.sol:194]: Done and added deposited collaterals returned as an array in `getPositionDigest`.

File: contracts/storage/Margin.sol (hunk 5 above)
Comment (davidvuong) [reply contracts/storage/Margin.sol:265]: Keeping it as it is for now.

**Review by bachstatter — APPROVED** [review]

**Pull request merged**